Adding PolygonArea for calculating the perimeter and area of a polygon on a Geodesic

phayes opened the pull request:
- [X] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [X] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.
---

This PR adds `PolygonArea` for calculating the perimeter and area of a polygon on a Geodesic. It is a port of the PolygonArea class in Geographic Lib (https://github.com/geographiclib/geographiclib/blob/main/src/PolygonArea.cpp). 

This PR also adds `accurate` as an optional dependency (enabled by default), that allows for very accurate summations. This replaces what would have gone in `accumulator.rs`

phayes committed:
Adding PolygonArea for calculating the perimeter and area of a polygon on a Geodesic

phayes committed:
Adding entry for PolygonArea to CHANGES.md

phayes committed:
Moving PolygonArea to the root of the crate

phayes committed:
Adding documentation to PolygonArea::new()

michaelkirk commented:
Do you know how this is being tested within the upstream library? Is there a test suite we could lift or seek inspiration from?

phayes commented:
@michaelkirk, I've gone looking for a test suite, and couldn't see anything obvious. I'll keep looking.  

At the very least I want to test it against a bunch of weird polygons and ensure that it has the same output as upstream.

phayes committed:
Adding  as an optional dependency

phayes committed:
Removing empty accumulator.rs, using  crate instead

michaelkirk commented:
Digging a bit, I haven't found a clear cpp test suite either, but I found some threads to chase down:

1. There is this cli tool, which I think is being used in a test harness for the cpp program's polygon area calculation:
https://github.com/geographiclib/geographiclib/blob/main/tools/Planimeter.cpp

2. A keyword to search for in the code base is "Planimeter" - I found a series of relevant looking tests in other languages:
https://github.com/geographiclib/geographiclib-octave/blob/main/inst/geographiclib_test.m#L644

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,202 @@
+        let perimetersum;
+        let areasum;
+
+        #[cfg(not(feature = "accurate"))]
```
Is it possible to do anything useful with PolygonArea without `accurate`? 

If not, maybe it's better to get rid of all the `accurate` feature checks sprinkled around, and add a `polygon_area` feature to this crate which *requires* accurate.

michaelkirk replied:
(sorry, that was an incomplete thought)

... the idea being that we'd then check for the `polygon_area` feature just once, in src/lib, and everything here could then infer we have it.

michaelkirk replied:
Sorry, I only just now realized that you're handling the case without `accurate` as well - please disregard my previous two comments. 🤦

phayes committed:
Making accurate enabled by default, adding winding and handling meridian transits

phayes committed:
Adding tests

phayes committed:
Adding docs to Winding enum

phayes committed:
Tests now pass

phayes committed:
Filling out test_planimeter6

phayes committed:
adding test_planimeter29

phayes commented:
Hi @michaelkirk , 

This is now ready for review.  I've added copies of all the tests I could find, and they are all now passing. 

Changes since the initial PR:

1. Adding `Winding` enum to make sure we are properly handling both clockwise and counterclockwise windings
2. `accurate` is now enabled by default. My thinking here is that if someone is making use of a geodesic area calculator, they're already motivated to have good accuracy over raw speed, so we should enable it by default. 
3. Improved and expanded documentation

phayes committed:
Removing dbg

phayes committed:
Improving readme

phayes committed:
Improving documentation

phayes committed:
Fixing typo

phayes committed:
Removing redundent comment

phayes committed:
Adding geodesic.area() and adding documentation on interpreting negative values

phayes commented:
Marked it as draft again as I found a bug in the `add_edge` method.  Fixing...

phayes commented:
Hi @michaelkirk ,

I might need a bit of help with this.  I'm chasing down this bug in `add_edge`, and I've been trying to make sure my understanding is correct.  I've come up with this test-case that is failing, but according to my own mind it should be passing.  Perhaps you can help explain where my understanding is confused?

```rust
    #[test]
    fn test_turnaround() {
        let g = Geodesic::wgs84();

        let start = (0.0, 0.0);
        let destination = (0.0, 1.0);

        let (distance, azi1, _, _) = g.inverse(start.0, start.1, destination.0, destination.1);

        // Confirm that we've gone due-east
        assert_eq!(azi1, 90.0);

        // Turn around by adding 180 degrees to the azimuth
        let turn_around = azi1 + 180.0;

        // Confirm that turn around is due west
        assert_eq!(turn_around, 270.0);

        // Test that we can turn around and get back to the starting point.
        let (lat, lon) = g.direct(destination.0, destination.1, turn_around, distance);
        assert_eq!(lat, start.0);
        assert_eq!(lon, start.1);
    }
```

phayes committed:
Fixing issue with transits on add_edge, adding more tests

phayes commented:
The bug in `add_edge` is now fixed. Ready for review!

phayes commented on Cargo.toml:
```diff
@@ -20,8 +20,11 @@ test_full = []
 # You must run script/download-test-data.sh before using this feature
 test_short = []
 
+default = ["accurate"]
```
Note that `accurate` is enabled by default

phayes committed:
Improve documentation

phayes committed:
Note addition of accurate feature in change log

michaelkirk commented on src/lib.rs:
```diff
@@ -1,3 +1,71 @@
+//! A subset of [geographiclib](https://geographiclib.sourceforge.io/) implemented in Rust.
```
I'm not really reviewing this hunk, since I'm assuming it's copy/pasted from the README. Let me know if I'm wrong.

phayes replied:
Yes correct, I added additional examples for PolygonArea and non-standard geoids (the example uses mars).

michaelkirk replied:
Yes I saw that - thanks for the new examples!

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,561 @@
+#[cfg(feature = "accurate")]
+use accurate::traits::*;
+
+/// Clockwise or Counterclockise winding
```
```suggestion
/// Clockwise or CounterClockise winding
```

michaelkirk commented on src/polygonarea.rs:
````diff
@@ -1 +1,561 @@
+    /// }
+    /// assert_eq!(area, 510053312945726.94);
+    /// ```
+    pub fn compute(mut self) -> (f64, f64) {
````
The upstream implementation of this method returns `self.num` and uses inout params for the "actual" results. Returning `self.num` makes it possible for the caller to tell if they've done something problematic, like called `compute` without first adding some points (if they check for it), so they can know that the results are garbage.

I'm in favor of returning a tuple, as you have, for the "results", since it's a small diversion from upstream and it's easy to do in rust, but I'd prefer to maintain a way for the caller to know that something went wrong and the results are garbage.

Generally I've been trying to stick pretty close to the upstream implementation, under the theory that it'll be easier to cross-reference with the upstream implementation in the future. Historically, all (or almost all) errors  in this crate have been simple transcription errors, so I think it's a good theory.

So: We could return a `(perimeter, area, num)` tuple to stay pretty close to the upstream implementation.

Alternatively, though it's a bigger divergence, I could be convinced to return something like `Ok((perimeter, area)) | Err(NotEnoughPointsError(num))`, which feels like more idiomatic rust.

What do you think @phayes?

phayes replied:
I've added `self.num` to the result tuple.

michaelkirk commented on src/polygonarea.rs:
````diff
@@ -1 +1,561 @@
+    ///     area += planetary_area;
+    /// }
+    /// assert_eq!(area, 510053312945726.94);
+    /// ```
````
Could you add a note to the effect of:

> Unlike upstream geographiclib, we do not support the `reverse` or `sign` parameters.  See the [upstream documentation](https://geographiclib.sourceforge.io/C++/doc/PolygonArea_8hpp_source.html) for details of what that means.

michaelkirk replied:
Oh! I see now - we actually __do__ support reverse, but it's via `self.winding`, which doesn't exist in upstream, right?

I think using a named Enum is better than using a bool, so I support that change. 

In the interest of aligning with upstream, how would you feel about moving `winding` from an instance variable to a parameter on this method?

It seems like it better aligns with upstream behavior with no downside that I can see, but feel free to push back.

phayes replied:
I think, given your desire to stick pretty close to upstream, I'll add the "sign" param.

phayes replied:
OK, the "sign" param has been added. I've updated the documentation to reflect this.

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,561 @@
+                .direct(self.latest_lat, self.latest_lon, azimuth, distance);
+        self.perimetersum += distance;
+        self.areasum += S12;
+        self.crossings += PolygonArea::transit(self.latest_lon, lon);
```
Upstream uses transitdirect here - why do we diverge? Is it safe?

phayes replied:
I removed `transitdirect` for two reasons:

1. My translation of `PolygonArea::transitdirect()` was buggy and causing tests to fail.  It's possible that upstream is buggy, I have additional tests for `add_edge` that don't exist in upstream.  Upstream is used for military applications though, so I assume it's VERY well tested, we just can't find the tests. Unsure.
2. We have all the data right at our fingertips to simply call `PolygonArea::transit()`.   In order for `Geodesic::direct()` to compute the area, it *has* to compute the next lat / lon pair. This means it's zero-cost to call `PolygonArea::transit()`.

phayes replied:
This is now fixed.  Using `transitdirect` as per upstream.

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,561 @@
+        let lon2 = ang_normalize(lon2);
+
+        // Translation from the following cpp code:
+        //  https://github.com/geographiclib/geographiclib/blob/main/src/PolygonArea.cpp#L22
```
Thank you for these cross references! Can you include a stable REF (like the commit sha or a tag) rather than `main`?

Same request for the other links in this PR.

phayes replied:
Done!

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,561 @@
+
+        let (perimeter, area) = pa.compute();
+
+        assert_relative_eq!(perimeter, 443770.917, epsilon = 1.0e-3);
```
It looks like these epsilon values were lifted from the upstream test suite, is that right?

phayes replied:
I lifted them from upstream, I also did some best effort "epsilon rationalization" to try to minimize the size of epsilon until the test failed, then backed up by one.

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,561 @@
+    }
+
+    #[test]
+    fn test_planimeter21() {
```
This looks to be planimeter19, not 21.
```octave
function n = Planimeter19
% Coverage tests, includes Planimeter19 - Planimeter20 (degenerate
% polygons).
  n = 0;
  points = [1,1];
  [area, perimeter] = geodarea(points(:,1), points(:,2));
  n = n + assertEquals(area, 0, 0);
  n = n + assertEquals(perimeter, 0, 0);
end
```

```suggestion
    fn test_planimeter19() {
```

Can you add the real planimeter21?

phayes replied:
I've added `test_planimeter21`, minus assrtions for `test_poin()` and `test_edge()` which we don't yet support.

phayes replied:
I've added more degeneracy tests to `test_planimeter19`. 

NOTE: I previously had a comment here about handling degenerate polygons.  Our implementation now matches upstream.

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,561 @@
+    }
+    
+    #[test]
+    fn test_planimeter29() {
```
I can't find a planimeter29 test case for octave, but I found this one for javascript. 

https://github.com/geographiclib/geographiclib-js/blob/57137fdcf4ba56718c64b909b00331754b6efceb/geodesic/test/geodesictest.js#L883

Is that the inspiration? Can you include a link?

phayes replied:
Done!

michaelkirk reviewed (changes requested):
Thanks for the PR @phayes! 

I have a lot of comments, but overwhelmingly I feel great about this PR. The only real concern I have is the divergence between upstream with the use of transit vs transitdirect in add_edge.

All the other comments are nits. If you don't have time or energy, I'm happy to merge without addressing the nits and can follow up later.

phayes committed:
Update src/polygonarea.rs

Co-authored-by: Michael Kirk <michael.code@endoftheworl.de>

phayes committed:
Returning number of points in the polygon

phayes committed:
Adding support for  param

phayes committed:
Fixing example code

phayes committed:
Using definite refs instead of main branch

phayes committed:
Adding more tests

phayes committed:
Adding source for test_planimeter29

phayes commented:
Hi @michaelkirk,

I've addressed the comments, the following decisions are outstanding:

1. Do we report the perimeter for degenerate polygons to be `NaN` or to be whatever naturally falls out of the algorithim? NOTE: This is now fixed and matches upstream.

2. Winding.  I really like the current pattern in the PR of setting the winding when creating the `PolygonArea`, it feels like a more natural API.  You define the winding, then you add points according to that winding.     However, feel free to push back on this. 

3. `transitdirect`.   My previous understanding was wrong. Working on a fix and more tests now.  NOTE: This is now fixed and matches upsteam.

phayes commented:
I'm gonna dig more into `transitdirect` and see if I can get more clairty

phayes committed:
Fixing add_edge and other misc issues due to wrong mask being used

phayes committed:
Adding missing assert from planimeter19

phayes commented:
Hi @michaelkirk ,

I believe I fixed all the outstanding issues.  `transitdirect` and degenerate polygons now work and behavior matches upstream.  The issue ended up being related to the mask being used. I was using the default mask, but `PolygonArea` needs to be using a specific mask. This is now fixed and everything has fallen into place. 

The only outstanding decision is that winding is specified during creation of the `PolygonArea`. Just need confirmation that we like this ergonomically.

phayes committed:
Fixing typo

phayes committed:
Adding more documentation to winding, also adding Default impl

phayes committed:
Making winding copy

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,646 @@
+#[cfg(feature = "accurate")]
+use accurate::traits::*;
+
+/// Clockwise or Counterclockwise winding
```
```suggestion
/// Clockwise or CounterClockwise winding
```

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -96,9 +98,9 @@ impl<'a> PolygonArea<'a> {
+            let (_a12, s12, _azi1, _azi2, _m12, _M12, _M21, S12) =
                 self.geoid
-                    .inverse(self.latest_lat, self.latest_lon, lat, lon);
+                    ._gen_inverse_azi(self.latest_lat, self.latest_lon, lat, lon, POLYGONAREA_MASK);
```
If you take a look at the implementations, you'll see that `_gen_inverse_azi` calls `_gen_inverse` and then does a little bit extra to compute `azi1`/`azi2`.

Since we're not using `azi1`/`azi2`, let's use `_gen_inverse` and save the unnecessary computation.

phayes replied:
Done!

michaelkirk commented on src/polygonarea.rs:
```diff
@@ -1 +1,646 @@
+    }
+
+    #[test]
+    fn test_edge_2() {
```
What are you testing here?

phayes replied:
Oops!  That was a halfway written test that I realized that I didn't need and forgot to delete it.   I've now deleted it.

michaelkirk reviewed (changes requested):
Good find on the output mask discrepancy!

re: WindingOrder - I'm fine to keep the API as you have it where winding order is declared at instantiation. Karney's API is more flexible in that you could compute the area from the perspective of one winding, and then the other, but I'm not sure that would even make sense to do. We can always revisit if somebody actually wants to do that in the future.

I left another round of feedback which I believe will be straight forward to address.

phayes committed:
Update src/polygonarea.rs

Co-authored-by: Michael Kirk <michael.code@endoftheworl.de>

phayes committed:
Fixing several issues as per comments

phayes committed:
Merge branch 'main' of github.com:phayes/geographiclib-rs

michaelkirk commented:
bors r+

Thanks so much!

michaelkirk commented:
Tests are failing with what looks to be a very small discrepancy. I'm comfortable merging it as is if you want to just bump the tolerance and merge yourself.

bors d=phayes

phayes committed:
Changing epsilon for a few failing tests

phayes commented:
That's super odd, the tests pass on my machine.   I wonder if there's subtle floating-point differences on Arm64 vs whatever is running on GitHub.

phayes commented:
bors try

phayes commented:
bors r+
